### PR TITLE
Support decompressor plugins

### PR DESCRIPTION
**Pull request by jspanjers**
As suggested in #425, this pull request makes it possible to easily extend rubyzip with additional decompressors.

### CODE REVIEW

File: lib/zip/inflater.rb [hunk 1]
```diff
@@ -1,64 +1,55 @@
...
 
     private
 
-    def internal_produce_input(buf = '')
+    def buffer
+      @buffer ||= ''.dup
+    end
+
+    def zlib_inflater
+      @zlib_inflater ||= ::Zlib::Inflate.new(-Zlib::MAX_WBITS)
```
[the remark this answers is not in the document]
Comment (jspanjers) [reply lib/zip/inflater.rb]: There is no really good reason I guess. Originally, I was trying to completely remove the initialize method, but there is no good reason for that either. I will adjust by patches to not use lazy initialization here.

File: lib/zip/inflater.rb [hunk 2]
```diff
@@ -1,64 +1,55 @@
...
       begin
-        @zlib_inflater.inflate(@decrypter.decrypt(@input_stream.read(Decompressor::CHUNK_SIZE, buf)))
+        zlib_inflater.inflate(input_stream.read(Decompressor::CHUNK_SIZE))
       rescue Zlib::BufError
         raise if retried >= 5 # how many times should we retry?
         retried += 1
         retry
       end
+    rescue Zlib::Error => e
+      raise(::Zip::DecompressionError, e.inspect)
```
[the remark this answers is not in the document]
Comment (jspanjers) [reply lib/zip/inflater.rb]: Agreed, I will make the adjustment.

File: lib/zip/input_stream.rb [hunk 3]
```diff
@@ -124,46 +120,54 @@ def get_io(io_or_file, offset = 0)
...
         raise ::Zip::CompressionMethodError,
               "Unsupported compression method #{@current_entry.compression_method}"
       end
+
+      decompressor_class.new(@decrypted_io, decompressed_size)
     end
 
     def produce_input
-      @decompressor.produce_input
+      @decompressor.read(CHUNK_SIZE)
```
[the remark this answers is not in the document]
Comment (jspanjers) [reply lib/zip/input_stream.rb]: I think the optimal CHUNK_SIZE for decompressors is dependant on compression method. Thus the Decompressor::CHUNK_SIZE should not exist and be replaced by an implementation dependant CHUCK_SIZE in Inflater.

Therefore, in InputStream, I think it is better to not depend on this Decompressor::CHUNK_SIZE.

**Comment by jspanjers** [comment]
> I think this could go out in a semver minor version bump, probably 2.2 (I am planning to release 2.1 shortly), because I don't see any breaking changes, except to classes marked with `:nodoc:`. Does that sound right?

Agreed.

> Are you planning on releasing a plugin gem with bzip2 compression support? You can of course keep that under your own namespace, but if you would like it to live under the rubyzip org, we can ask @simonoff about setting that up.

Yes, I am planning to create a bzip2 compression plugin gem, after this pull request is accepted.  When the plugin gem is ready, I would very much like to transfer control (and maintainership) to rubyzip.org.

**Comment by hainesr** [comment]
>> Are you planning on releasing a plugin gem with bzip2 compression support? You can of course keep that under your own namespace, but if you would like it to live under the rubyzip org, we can ask @simonoff about setting that up.

> Yes, I am planning to create a bzip2 compression plugin gem, after this pull request is accepted. When the plugin gem is ready, I would very much like to transfer control (and maintainership) to rubyzip.org.

I think developing a collection of plugins under the rubyzip org is a nice idea. It would also be an opportunity to widen the cohort of maintainers of the library and drive more participation.

**Comment by Kimmie234** [comment]
Wallet

**Pull request merged**